packages/guides/src/Compiler/NodeTransformers/SectionCreationTransformer.php

```diff
@@ -0,0 +1,112 @@
+use phpDocumentor\Guides\Nodes\DocumentNode;
+use phpDocumentor\Guides\Nodes\Node;
+use phpDocumentor\Guides\Nodes\SectionNode;
+use phpDocumentor\Guides\Nodes\TitleNode;
+
+use function array_pop;
+use function count;
+use function end;
+
+use const PHP_INT_MAX;
+
+/** @implements NodeTransformer<Node> */
+final class SectionCreationTransformer implements NodeTransformer
+{
+    /** @var SectionNode[] $sectionStack */
+    private array $sectionStack = [];
+
+    public function enterNode(Node $node, CompilerContext $compilerContext): Node
+    {
+        if (!$compilerContext->getShadowTree()->getParent()?->getNode() instanceof DocumentNode) {
+            return $node;
+        }
+
+        if (!$node instanceof TitleNode) {
+            $lastSection = end($this->sectionStack);
+            if ($lastSection instanceof SectionNode) {
+                $lastSection->addChildNode($node);
+            }
+        }
+
+        return $node;
+    }
+
+    public function leaveNode(Node $node, CompilerContext $compilerContext): Node|null
+    {
+        if (!$compilerContext->getShadowTree()->getParent()?->getNode() instanceof DocumentNode) {
+            return $node;
+        }
+
+        if ($node instanceof SectionNode) {
+            return $node;
+        }
+
+        if (count($this->sectionStack) === 0 && !$node instanceof TitleNode) {
+            return $node;
+        }
+
+        if (count($this->sectionStack) > 0 && $compilerContext->getShadowTree()->isLastChildOfParent()) {
+            $lastSection = end($this->sectionStack);
+            while ($lastSection?->getTitle()->getLevel() > 1) {
+                $lastSection = array_pop($this->sectionStack);
+            }
+
+            return $lastSection;
+        }
+
+        if ($node instanceof TitleNode) {
+            $lastSection = end($this->sectionStack);
+            if ($lastSection instanceof SectionNode && $node !== $lastSection->getTitle() && $node->getLevel() <= $lastSection->getTitle()->getLevel()) {
+                while (end($this->sectionStack) instanceof SectionNode && $node !== end($this->sectionStack)->getTitle() && $node->getLevel() <= end($this->sectionStack)->getTitle()->getLevel()) {
+                    $lastSection = array_pop($this->sectionStack);
+                }
+
+                $newSection = new SectionNode($node);
+                if (end($this->sectionStack) instanceof SectionNode) {
+                    end($this->sectionStack)->addChildNode($newSection);
+                }
+
+                $this->sectionStack[] = $newSection;
+
+                return $lastSection?->getTitle()->getLevel() === 1 ? $lastSection : null;
+            } else {
+                $newSection = new SectionNode($node);
+                if ($lastSection instanceof SectionNode) {
+                    $lastSection->addChildNode($newSection);
+                }
+
+                $this->sectionStack[] = $newSection;
+            }
```

Comment:
As the previous if block is already returning, this does not need to be an `else`
```suggestion
              $newSection = new SectionNode($node);
              if ($lastSection instanceof SectionNode) {
                  $lastSection->addChildNode($newSection);
              }

              $this->sectionStack[] = $newSection;
            
```